src/main/scala/swiss/dasch/domain/BulkIngestService.scala

```diff
@@ -38,22 +38,26 @@ final case class BulkIngestService(
       _         <- semaphoresPerProject.put(key, semaphore)
     } yield semaphore).commit
 
-  private def acquireWithTimeout(sem: TSemaphore): UIO[Unit] =
-    sem.acquire.commit.timeout(Duration.fromMillis(400)).as(())
+  private def acquireWithTimeout(sem: TSemaphore): IO[Unit, Unit] =
+    sem.acquire.commit.timeout(Duration.fromMillis(400)).some.mapError(_ => ())
 
-  private def withSemaphore[E, A](key: ProjectShortcode)(
+  private def withSemaphoreDaemon[E, A](key: ProjectShortcode)(
     zio: IO[E, A],
-  ): IO[Option[E], A] =
+  ): IO[Unit, Fiber.Runtime[E, A]] =
     getSemaphore(key)
       .tap(acquireWithTimeout(_).asSomeError)
-      .flatMap(sem => zio.logError.ensuring(sem.release.commit).asSomeError)
+      .flatMap(sem => zio.logError.ensuring(sem.release.commit).forkDaemon)
+      .mapError(_ => ())
 
-  def startBulkIngest(shortcode: ProjectShortcode): IO[Option[Nothing], Fiber.Runtime[IOException, IngestResult]] =
-    withSemaphore(shortcode) {
-      doBulkIngest(shortcode).forkDaemon
+  private def withSemaphore[E, A](key: ProjectShortcode)(zio: IO[E, A]): IO[Option[E], A] =
```

Comment:
The error channels for the both `withSemaphore*` functions are ok and model the states we could have. However, I find it awkward that the case when a project is locked is modelled with two completely different types: `Unit` and `None`.

How about introducing a `case object ProjectLocked` and use it as the return value. With Scala 3 union type this would work quite nice, like so:

* `withSemaphoreDaemon` returns `IO[ProjectLocked.type, Fiber.Runtime[E, A]]`
* `withSemaphore` returns `IO[ProjectLocked.type | E, A]`

Or if you think this is overkill we could replace the `Unit` with `Option[Nothing]`?

Reply:
I think it is appropriate, given that in one of the cases, there's no alternative to Unit. It doesn't serve any particular purpose other than technically aligning the types to fit the feeling that these should be united.

`ProjectLocked.type | E` would technically be the solution to make it really precise, however, it also robs us of any Option methods like getOrElse.

I'll take a brief look at this again on Monday, and I can implement this at your request if necessary, but I do think it's fine as it is.

Reply:
Agreed to close as is for now.